### PR TITLE
Added SPOSiteDesign resource

### LINKED ISSUE #69: TeamsChannel Timing Issues
#### Details of the scenario you tried and the problem that is occurring
If in the same config, you create a Team and then try to add a channel to it, the TeamsChannel resource fails and complains that the team doesn't exist. It seems to be a timing issue because if you re-run the same config a few seconds later then everything works fine. 

#### Verbose logs showing the problem
PowerShell DSC resource MSFT_TeamsChannel  failed to execute Test-TargetResource functionality with error message: Team with Name Collab Summit
doesnt exist in tenant

#### Suggested solution to the issue
Add a few retries in the TeamsChannel resource whenever the Teams is not found. Maybe just 6 times 5 seconds.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
TeamsTeam CollabSummit
        {
            DisplayName        = "Collab Summit"
            Description        = "This is me demoing the Teams Resource"
            GlobalAdminAccount = $GlobalAdmin
            Ensure             = "Present"
        }

        TeamsChannel DSCChannel
        {
            TeamName           = "Collab Summit"
            DisplayName        = "DSC Discussions"
            GlobalAdminAccount = $GlobalAdmin
            Ensure             = "Present"
        }
```

#### The operating system the target node is running
NA

### PR DESCRIPTION
**Pull request by desmay**
#### Pull Request (PR) description
New SPOSiteDesign resource.
#### This Pull Request (PR) fixes the following issues
- New SPOSiteDesign resource
- Fixes #69  

Still need to add Teams timing fix to remaining Teams resources

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/office365dsc/74)

### CODE REVIEW

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 [hunk 1]
```diff
@@ -953,7 +953,38 @@ function Get-TeamByGroupID
...
+        $loopCounter +=1
+        if ($loopCounter -gt 5)
+        {
+            break
+        }
+    } while ($null -eq $team)
 
+    if ($null -eq $team)
+    {
+        throw "Team with Name $TeamName doesnt exist in tenant"
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: add apostrophe to doesn't

File: Tests/Unit/Stubs/Office365.psm1 [hunk 2]
```diff
@@ -11487,6 +11495,50 @@ function Grant-PnPSiteDesignRights
...
+        [System.String]
+        $WebTemplate,
+
+        [Parameter()]
+        [System.String[]]
+        $SiteScriptIds,
+
+        [Parameter()]
+        [System.Boolean]
+        $isDefault,
```
Comment (NikCharlebois) [inline Tests/Unit/Stubs/Office365.psm1]: IsDefault

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        SiteDesign SiteDesignConfig {
+            Title               = "DSC Site Design"
+            SiteScriptNames     = "Cust List", "List_Views"
+            WebTemplate         = "TeamSite"
+            isDefault           = $false
+            Description         = "Created by DSC"
+            PreviewImageAltText = "Office 365"
+            PreviewImageUrl     = ""
+            Ensure              = "Present"
+            Version             = 1
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md]: Just major

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        return $nullReturn
+    }
+
+    $scriptTitles = @()
+    foreach ($scriptId in $siteDesign.SiteScriptIds)
+    {
+        $siteScript = Get-PnPSiteScript -Identity $scriptId
+        $scriptTitles += $siteScript.Title
+    }
+    ## Todo need to see if we can get this PBP module instead of hard coded in scipt
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: done

[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    $CurrentParameters = $PSBoundParameters
+    $CurrentParameters.Remove("CentralAdminUrl")
+    $CurrentParameters.Remove("GlobalAdminAccount")
+    $CurrentParameters.Remove("SiteScriptNames")
+    $CurrentParameters.Remove("Ensure")
+    $CurrentParameters.Add("SiteScriptIds", $scriptIds)
+
+    if ($curSiteDesign.Ensure -eq "Absent")
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: Done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    {
+        $CurrentParameters.Remove("Version")
+        Write-Verbose -Message "Adding new site design $Title"
+        Add-PnPSiteDesign @CurrentParameters
+    }
+    ## Found current site design needs updated
+    if ($curSiteDesign.Ensure -eq "Present" -and "Present" -eq $Ensure)
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: Done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    {
+        $siteDesign = Get-PnPSiteDesign -Identity $Title -ErrorAction SilentlyContinue
+        if ($null -ne $siteDesign)
+        {
+            Write-Verbose -Message "Updating current site design $Title"
+            Set-PnPSiteDesign  -Identity $siteDesign.Id  @CurrentParameters
+        }
+    }
+
+    if ($Ensure -eq "Absent")
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    {
+        $siteDesign = Get-PnPSiteDesign -Identity $Title -ErrorAction SilentlyContinue
+        if ($null -ne $siteDesign)
+        {
+            Write-Verbose -Message "Removing site design  $Title"
+            Remove-PnPSiteDesign -Identity $siteDesign.Id  -Force
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof [hunk 9]
```diff
@@ -0,0 +1,16 @@
+[ClassVersion("1.0.0.0"), FriendlyName("SPOSiteDesign")]
+class MSFT_SPOSiteDesign : OMI_BaseResource
+{
+    [Key, Description("The title of the site design")] string Title;
+    [Write, Description("The names of the site design scripts")] string SiteScriptNames[];
+    [Write, Description("Web template to which site design is applied to when invoked."), ValueMap{"CommunicationSite","TeamSite"}, Values{"CommunicationSite","TeamSite"}] string WebTemplate;
+    [Write, Description("Description of site design")] string Description;
+    [Write, Description("Is site design applied by default to web templates .")] Boolean isDefault;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    [Write, Description("Site design alt preview image text")] string PreviewImageAltText;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    WebTemplate
+      - Required: No
+      - Description: Web template to apply site design, only validate
+        parameters are CommunicationSite and TeamSite
+
+    Description
+      - Required: No
+      - Description: Description of the site design
+
+    isDefault
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    GlobalAdminAccount
+      - Required: Yes
+      - Description: Credentials of the Office365 Tenant Admin
+
+## Example
+
+```PowerShell
+        SiteDesign SiteDesignConfig {
+            Title               = "DSC Site Design"
+            SiteScriptNames     = "Cust List", "List_Views"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            WebTemplate         = "TeamSite"
+            isDefault           = $false
+            Description         = "Created by DSC"
+            PreviewImageAltText = "Office 365"
+            PreviewImageUrl     = ""
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/readme.md]: done

File: Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1 [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+Configuration SiteDesignConfig
+{
+    Import-DSCResource -ModuleName Office365DSC
+    $credsGlobalAdmin = Get-Credential -UserName "TenantAdmin@O365DSC1.onmicrosoft.com" -Message "Global Admin"
+    Node localhost
+    {
+        SPOSiteDesign SiteDesign
+        {
+            Title               = "DSC Site Design"
+            SiteScriptNames     = "Cust List", "List_Views"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1]: done

File: Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1 [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+            WebTemplate         = "TeamSite"
+            isDefault           = $false
+            Description         = "Created by DSC"
+            PreviewImageAltText = "Office 365"
+            PreviewImageUrl     = ""
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1]: done

File: Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1 [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    Import-DSCResource -ModuleName Office365DSC
+    $credsGlobalAdmin = Get-Credential -UserName "TenantAdmin@O365DSC1.onmicrosoft.com" -Message "Global Admin"
+    Node localhost
+    {
+        SPOSiteDesign SiteDesign
+        {
+            Title               = "DSC Site Design"
+            SiteScriptNames     = "Cust List", "List_Views"
+            WebTemplate         = "TeamSite"
+            isDefault           = $false
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.ps1]: done

File: Tests/Unit/Stubs/Office365.psm1 (hunk 2 above)
Comment (desmay) [reply Tests/Unit/Stubs/Office365.psm1]: done

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 1 above)
Comment (desmay) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof [hunk 17]
```diff
@@ -0,0 +1,16 @@
+[ClassVersion("1.0.0.0"), FriendlyName("SPOSiteDesign")]
+class MSFT_SPOSiteDesign : OMI_BaseResource
+{
+    [Key, Description("The title of the site design")] string Title;
+    [Write, Description("The names of the site design scripts")] string SiteScriptNames[];
+    [Write, Description("Web template to which site design is applied to when invoked."), ValueMap{"CommunicationSite","TeamSite"}, Values{"CommunicationSite","TeamSite"}] string WebTemplate;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.schema.mof]: done

File: Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1 [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+            Set-PnPSiteDesign  -Identity $siteDesign.Id  @CurrentParameters
+        }
+    }
+
+    if ($Ensure -eq "Absent")
+    {
+        $siteDesign = Get-PnPSiteDesign -Identity $Title -ErrorAction SilentlyContinue
+        if ($null -ne $siteDesign)
+        {
+            Write-Verbose -Message "Removing site design  $Title"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_SPOSiteDesign/MSFT_SPOSiteDesign.psm1]: done

File: Tests/Unit/Stubs/Office365.psm1 (hunk 2 above)
Comment (NikCharlebois) [reply Tests/Unit/Stubs/Office365.psm1]: @desmay I still see it lower case

File: Tests/Unit/Stubs/Office365.psm1 [hunk 19]
```diff
@@ -11511,6 +11563,58 @@ function Get-PnPSiteDesign
...
+        [string]
+        $Identity,
+
+        [Parameter()]
+        [string]
+        $WebTemplate,
+
+        [Parameter()]
+        [System.Boolean]
+        $isDefault,
```
Comment (NikCharlebois) [inline Tests/Unit/Stubs/Office365.psm1]: Still Lower Case

Comment (desmay) [reply Tests/Unit/Stubs/Office365.psm1]: sorry missed those I think I got them all now

**Comment by NikCharlebois** [comment]
:lgtm:
:shipit:

**Pull request merged**